tests/0.2.0.json

```diff
@@ -391,7 +381,246 @@
         "logical": {
           "type": "string"
         }
+      },
+      "not": {
+        "required": [
+          "alias"
+        ]
       }
+    },
+    "RecapLogicalTypes": {
+      "allOf": [
+        {
+          "if": {
```

Comment:
TIL about [conditionals in jsonschema](https://json-schema.org/understanding-json-schema/reference/conditionals.html#if-then-else). that's cool! 

I was expecting this to be implemented with `oneOf` and a discriminated union like the other RecapTypes, but I suppose doing it this way allows for non builtin logical types

Reply:
Haha yea.. that was me a couple days ago :P

Reply:
> I was expecting this to be implemented with oneOf and a discriminated union like the other RecapTypes, but I suppose doing it this way allows for non builtin logical types

Do you think the oneOf way is clearer? What would the "dicriminating union" look like in that case? I'm not totally sold on the if/then.. it feels code-y to me.

Reply:
```
    "RecapType": {
      "oneOf": [
        {
          "$ref": "#/$defs/RecapNull"
        },
        {
          "$ref": "#/$defs/RecapBool"
        },
```
 `RecapType` is `oneOf` `RecapNull`, ~RecapBool`, ... - where `Recap*` looks like
 
```
     "RecapNull": {
      "type": "object",
      "required": ["type"],
      "properties": {
        "type": {
          "type": "string",
          "enum": ["null"]
        },
```

The _union_ is the `oneOf`, and the system _discriminates_ between them based on the `type` field.

Since logical types appear to be subsets of types, you might be able to union the existing `RecapBytes` with the decimal-bytes logical type? I'm not 100% clear on what logical types are and how they function in recap though...



Also, using oneOf vs if-then might yield different error validation outputs that may be more or less user friendly, so that may be a consideration when designing the json schema meta schema.

Reply:
I'll poke at it a bit tomorrow, timeboxed to an hour or two. If I can't get something definitively better, I'll move on. Can always change it later.

Reply:
> Would you accept a PR changing this format?

Yes! But I think we need to be careful about making sure that the spec JSON schema continues to express the actual rules. I'm concerned you might find you need to relax the JSON schema in order to make it work with the generator. Doing so would no longer make it a valid spec.

> I'd also need some guidance on how to test it since the tests [hard code](https://github.com/recap-build/recap/blob/main/tests/spec/test_json_schema_spec.py#L23) the published schema.

IIRC, you added [an environment variable](https://github.com/recap-build/recap/blob/main/tests/spec/test_json_schema_spec.py#L28C38-L28C59) for such an occasion... :)

Reply:
WDYT about:

1. Adding the `generator` task to the CI (it should fail at first)
2. Updating the JSON schema (without updating any tests)
3. Validating that the generator now works

I think if  you can do this without having to modify the tests, it *should* be a safe spec change... We'll still have to sniff around since the tests don't cover every single edge case. But it should be pretty close...

Reply:
@mjperrone checking in here

Reply:
sorry, other priorities came up. I think I have time to at least make the initial PR changing the definition method, not sure I'll have the time to validate it

Reply:
[github.com/recap-build/recap-website/pull/17](https://github.com/recap-build/recap-website/pull/17): `use discriminated union (oneOf) instead of if/then to define logical types` (+156, −209)